core/src/main/scala/org/apache/spark/MapOutputTracker.scala

```diff
@@ -519,17 +521,19 @@ private[spark] abstract class MapOutputTracker(conf: SparkConf) extends Logging
    * but endMapIndex is excluded). If endMapIndex=Int.MaxValue, the actual endMapIndex will be
    * changed to the length of total map outputs.
    *
-   * @return A sequence of 2-item tuples, where the first item in the tuple is a BlockManagerId,
-   *         and the second item is a sequence of (shuffle block id, shuffle block size, map index)
-   *         tuples describing the shuffle blocks that are stored at that block manager.
-   *         Note that zero-sized blocks are excluded in the result.
+   * @return A case class object which includes two attributes. The first attribute is a sequence
+   *         of 2-item tuples, where the first item in the tuple is a BlockManagerId, and the
+   *         second item is a sequence of (shuffle block id, shuffle block size, map index) tuples
+   *         tuples describing the shuffle blocks that are stored at that block manager. Note that
+   *         zero-sized blocks are excluded in the result. The second attribute is a boolean flag,
+   *         indicating whether batch fetch can be enabled.
    */
   def getMapSizesByExecutorId(
       shuffleId: Int,
       startMapIndex: Int,
       endMapIndex: Int,
       startPartition: Int,
-      endPartition: Int): Iterator[(BlockManagerId, Seq[(BlockId, Long, Int)])]
+      endPartition: Int): MapSizesByExecutorId
```

Comment:
This changed the return type, which's supposed to be a part of API. ShuffleManager relies on this could still be broken.

Reply:
`MapOutputTracker` is `private[spark]` and not part of exposed API.
Am I missing something here ?

Reply:
Yeah, but as you may know, it's very common that Spark private APIs are abused as public APIs. We did break downstream use cases by changing private APIs, e.g., SQL UDF stuff.  And I know some users implement their own shuffle module by extending `ShuffleManager`,  `MapOutputTracker`, etc. Given that we're very close to the 3.2 release, I think it's not appropriate to have this kind of change at this moment. 

If you guys still insist on  this way, I'd suggest to totally separate the push-based shuffle path with the normal shuffle path from `SortShuffleManager.getReader`, e.g.,

```scala
  override def getReader[K, C](
      handle: ShuffleHandle,
      startMapIndex: Int,
      endMapIndex: Int,
      startPartition: Int,
      endPartition: Int,
      context: TaskContext,
      metrics: ShuffleReadMetricsReporter): ShuffleReader[K, C] = {
    val (blocksByAddress, enableBatchFetch ) = if (fetchMergeResult) {
      val res = SparkEnv.get.mapOutputTracker.getPushBasedShuffleMapSizesByExecutorId(...)
      (res.iter, res.enableBatchFetch)
    } else {
      val address = SparkEnv.get.mapOutputTracker.getMapSizesByExecutorId(...)
      (address, canUseBatchFetch(startPartition, endPartition, context))
    }
    val blocksByAddress = 
    new BlockStoreShuffleReader(
      handle.asInstanceOf[BaseShuffleHandle[K, _, C]], blocksByAddress., context, metrics,
      shouldBatchFetch = enableBatchFetch)
          
  }

```

Reply:
Given how close we are to RC, I am fine with this approach.

In general though, we should be very careful about setting expectation that there would be compatibility guarantees with `private[spark]` classes; they are explicitly marked that way to make it very clear not to depend on them. Inspite of that, if there are projects/users depending on it, it is up to them to ensure compatibility - not spark project.

@zhouyejoe Can you evaluate the change that @Ngone51 proposed please ? Internally, it could delegate to the same `getMapSizesByExecutorIdImpl` (which would be what you have here) and simply return response.iter for existing method.

Reply:
Ok. Will try it out tonight!

Reply:
@zhouyejoe btw, could you also update the doc of `spark.shuffle.push.enabled` to reflect the change made in this PR, e.g., "push-based shuffle takes priority over batch fetch for the scenario like partition coalesce."

Reply:
Didn't get enough time to work it tonight. Will update the PR tomorrow ASAP.

Reply:
Maybe, an equivalent way to do the same thing by `MapSizesByExecutorId` is:

```scala
override def getReader[K, C](
      handle: ShuffleHandle,
      startMapIndex: Int,
      endMapIndex: Int,
      startPartition: Int,
      endPartition: Int,
      context: TaskContext,
      metrics: ShuffleReadMetricsReporter): ShuffleReader[K, C] = {
    val blocksByAddress = SparkEnv.get.mapOutputTracker.getMapSizesByExecutorId(
      handle.shuffleId, startMapIndex, endMapIndex, startPartition, endPartition)
   val hasMergedBlock = blocksByAddress.exists(_._1.host == SHUFFLE_MERGER_IDENTIFIER)
    new BlockStoreShuffleReader(
      handle.asInstanceOf[BaseShuffleHandle[K, _, C]], blocksByAddress, context, metrics,
      shouldBatchFetch =  !hasMergedBlock && canUseBatchFetch(startPartition, endPartition, context))
  }
```
which looks simpler.

Reply:
@Ngone51 We would be trying to infer behavior in this case based on the response, instead of being explicit about what should be enabled/disabled (btw, `blocksByAddress` is iterator, so we would need a `.toList` or some such as well).

The implication of that is, for example, without the `mergeStatuses.exists(_.exists(_ != null))` in `convertMapStatuses`, `hasMergedBlock` will be false - and so we would enable batch fetch - and fail.
I want to avoid tight coupling based on pattern of blocks in response, and instead explicitly let the response indicate whether we should enable/disable feature. 

Thoughts ?

Reply:
Sounds good.